MM2RandoLib/Randomizers/RText.cs

```diff
@@ -1,226 +1,238 @@
+                Int32 characterIndex = 0;
+                foreach (Char c in weaponName)
                 {
-                    if (j < chars.Length)
-                    {
-                        byte b = Convert.ToByte(chars[j]);
-                        p.Add(offset + j, b, String.Format("Weapon Name {0} Char #{1}: {2}", ((EDmgVsBoss.Offset)i).Name, j, chars[j].ToString()));
+                    in_Patch.Add(
+                        offsetAddress + characterIndex,
+                        c.AsPrintCharacter(),
+                        String.Format("Weapon Name {0} Char #{1}: {2}", ((EDmgVsBoss.Offset)weaponIndex).Name, characterIndex, c.ToString()));
 
-                    }
-                    else
-                    {
-                        p.Add(offset + j, Convert.ToByte('@'), $"Weapon Name {((EDmgVsBoss.Offset)i).Name} Char #{j}: @");
-                    }
+                    characterIndex++;
                 }
             }
 
             // Erase "Boomerang" for now
-            for (int i = 0; i < 10; i++)
+            // TODO: why?
+            for (Int32 i = 0; i < 10; i++)
             {
-                p.Add(0x037f5e + i, Convert.ToByte('@'), $"Quick Boomerang Name Erase Char #{i}: @");
+                in_Patch.Add(0x037f5e + i, ' '.AsPrintCharacter(), $"Quick Boomerang Name Erase Char #{i}: @");
             }
 
-            // Create new weapon letters
-            {
-                // Keep local copy of the alphabet to remove letters from
-                List<char> alphabet = new List<char>(Alphabet.ToCharArray());
-
-                // First pick a letter for buster, 1/26
-                int rLetterIndex = r.Next(alphabet.Count);
-                newWeaponLetters[0] = alphabet[rLetterIndex];
-                alphabet.RemoveAt(rLetterIndex);
+            // Pick a new weapon letter for buster
+            Char? busterWeaponLetter = WeaponName.GetUnusedWeaponLetter(in_Random, weaponNames);
 
-                // For each special weapon...
-                for (int i = 0; i < 8; i++)
-                {
-                    // Try to use the first letter of the weapon name if it hasn't been used yet
-                    char tryLetter = newWeaponNames[i][0];
-                    if (alphabet.Contains(tryLetter))
-                    {
-                        newWeaponLetters[i + 1] = tryLetter;
-                        alphabet.Remove(tryLetter);
-                    }
-                    // Otherwise use a random letter from the remaining letters
-                    else
-                    {
-                        rLetterIndex = r.Next(alphabet.Count);
-                        newWeaponLetters[i + 1] = alphabet[rLetterIndex];
-                        alphabet.RemoveAt(rLetterIndex);
-                    }
-                }
-            }
+            List<Char> weaponLetters = weaponNames.Select(x => x.Letter).ToList();
+            weaponLetters.Insert(0, busterWeaponLetter ?? 'P');
```

Comment:
So is `Char?` a sum type of a `Char` and distinguished constructor that acts like a `Null` value? and `??` provides the `Char` when it's a `Char` and the thing on right when it's the distinguished value?